### ISSUE TITLE
Unpermissioned users visiting a PKC where they can read a page, can't verify it.

### ISSUE THREAD

**Issue by FantasticoFox** [inblockio/DataAccounting#187]
![image](https://user-images.githubusercontent.com/45313235/143879187-02530f79-3eea-47e4-9aba-a6cc9ef901ef.png)

Users which can read a page, should be able to verify it. How to change the code responsible for permissions so that READ API's are open for the whitelisted pages in $wgWhitelistRead

**Comment by FantasticoFox** [comment]
I'm pretty sure it's related to the way how we use the permission manager in the API handlers. Maybe that helps?

**Comment by FantasticoFox** [comment]
Known Limitation for v1.0 but is critical for working with PKC for Publishing.

**Comment by pwirth** [comment]
ive left a wip on this issue:
https://github.com/inblockio/DataAccounting/commit/2c00d690bb3e2d00eee692c12ef6b2eacf1fddf0
in branch:
https://github.com/inblockio/DataAccounting/commits/restauth

This is not an easy fix, as MediaWiki rest api can not check for the $wgWhitelistRead as it does not know on which title this operation is for. Unfortunately a lot of stuff in the api is final or private. That means, the only way i found is to knockout the BasicAuthorizerInterface und apply own permission checks on the title, that is somehow provided from the prameters

**Comment by pwirth** [comment]
These rest routes are now permission checked to the page calculated from params:
* /data_accounting/verify_page/{rev_id}
* /data_accounting/get_page_by_rev_id/{rev_id}
* /data_accounting/get_page_all_revs/{page_title}
* /data_accounting/get_page_last_rev/{page_title}
* /data_accounting/request_hash/{rev_id}

These rest routes are now not permission checked at all. This means, a user doeas not even need global read permission to request date here!
* /data_accounting/request_merkle_proof/{witness_event_id}/{revision_verification_hash}
* /data_accounting/get_witness_data/{witness_event_id}
* /data_accounting/get_server_info

The routes i have not touched (yet):
* /data_accounting/write/store_signed_tx
* /data_accounting/write/store_witness_tx

Will these have to be opened as well and checked against the page?

**Comment by rht** [comment]
I think the write APIs can be excluded from this issue requirement? @FantasticoFox do you want random strangers to be able to sign a whitelisted page?

**Comment by rht** [comment]
I want to list other possible solutions to the problem:
- Users could maintain 2 separate PKCs. One is completely public, the other is private. No new implementation is needed, and if the private PKC is air-gapped, the users can be assured of no sensitive data leakage. If the users want to expose a page, it simply exports the said page via XML to the public PKC
- We build another API endpoint that is similar to an XML import. So you just need to specify a page title, and it will return all the required data for you to verify the verification data
- We could ask the external verifier to use the export API to get the verification data. I assume that whitelisted pages can always be exported by anon users?

Also, I'm concerned of the timing attack that could expose whether the PKC server contains a particular page title, or the number of revisions it has

**Comment by FantasticoFox** [comment]
> I think the write APIs can be excluded from this issue requirement? @FantasticoFox do you want random strangers to be able to sign a whitelisted page?

This is correct. No Write API needed.

**Comment by FantasticoFox** [comment]
````
These rest routes are now not permission checked at all. This means, a user does not even need global read permission to request data here!

/data_accounting/request_merkle_proof/{witness_event_id}/{revision_verification_hash}
/data_accounting/get_witness_data/{witness_event_id}
/data_accounting/get_server_info
````
Currently requesting the witness id #1... and counting up integer will return something. This will change by us using the witness_hash to request the witness event. This will make it practically impossible to guess the witness event id.

That there is no read protection is fine, as there is no meaningful data leakage by my current understanding.

**Comment by FantasticoFox** [comment]
> I want to list other possible solutions to the problem:
> 
> * Users could maintain 2 separate PKCs. One is completely public, the other is private. No new implementation is needed, and if the private PKC is air-gapped, the users can be assured of no sensitive data leakage. If the users want to expose a page, it simply exports the said page via XML to the public PKC
> * We build another API endpoint that is similar to an XML import. So you just need to specify a page title, and it will return all the required data for you to verify the verification data
> * We could ask the external verifier to use the export API to get the verification data. I assume that whitelisted pages can always be exported by anon users?
> 
> Also, I'm concerned of the timing attack that could expose whether the PKC server contains a particular page title, or the number of revisions it has

I would like to discuss this point by point in person, as this relates to other issues down the road related to the federated PKC's. Whats the best way to export / import data. Maybe an clean Import / Export API might be the solution. Lets do a design session on this.

**Comment by pwirth** [comment]
oke, tests all green now :)

**Comment by rht** [comment]
Closed via #210.
